auction-server/src/opportunity/repository/add_opportunity.rs

```diff
@@ -63,4 +70,42 @@ impl<T: InMemoryStore> Repository<T> {
 
         Ok(opportunity)
     }
+
+    async fn check_db_duplicate_opportunity(
+        &self,
+        db: &sqlx::Pool<Postgres>,
+        opportunity: &<T as InMemoryStore>::Opportunity,
+    ) -> Result<bool, RestError> {
+        let chain_type = <T::Opportunity as entities::Opportunity>::ModelMetadata::get_chain_type();
+        let metadata = opportunity.get_models_metadata();
+
+        let result = sqlx::query!("SELECT COUNT(*) FROM opportunity WHERE permission_key = $1 AND chain_id = $2 AND chain_type = $3 AND sell_tokens = $4 AND buy_tokens = $5 AND metadata #- '{slot}' = $6",
+        opportunity.permission_key.to_vec(),
+        opportunity.chain_id,
+        chain_type as _,
+        serde_json::to_value(&opportunity.sell_tokens).unwrap(),
+        serde_json::to_value(&opportunity.buy_tokens).unwrap(),
+        self.remove_slot_field(serde_json::to_value(metadata).expect("Failed to serialize metadata")))
+            .fetch_one(db)
+            .await
+            .map_err(|e| {
+                tracing::error!("DB: Failed to check duplicate opportunity: {}", e);
+                RestError::TemporarilyUnavailable
+            })?;
+
+        match result.count {
+            Some(count) => Ok(count > 0),
+            None => {
+                tracing::error!("DB: Failed to check duplicate opportunity: count is None");
+                Err(RestError::TemporarilyUnavailable)
+            }
+        }
+    }
+
+    fn remove_slot_field(&self, mut metadata: serde_json::Value) -> serde_json::Value {
```

Comment:
Don't remove the slot. It's kind of metadata inside the db and we may need it for future tracing